### DISCUSSION TITLE
Proposal: SelectionModel should allow two state and three state inner node values

### DISCUSSION THREAD

**Discussion by grokys** [microsoft/microsoft-ui-xaml#7502] [General]
# Discussion: SelectionModel and TreeViews

I notice that `SelectionModel`'s support for hierarchical data is based on a "checkbox" model where a non-leaf node is considered selected if all of its children are selected, that is:

`IsSelected(0) == null` when some but not all children are selected

```
[-] Root
  [x] Child
  [ ] Child
```

`IsSelected(0) == true` when all children are selected

```
[x] Root
  [x] Child
  [x] Child
```

Note that in this example, "Root" was never selected explicitly.

This is great for a tree of checkboxes, but this isn't how a multi-selectable `TreeView` works, where a user must select each node with a click etc.

I notice that WinUI's `TreeView` doesn't use `SelectionModel`: is that because `SelectionModel` isn't designed for this use-case? Should it be?

**Comment by ranjeshj** [comment]
@grokys The current implementation of TreeView in WinUI was built before ItemsRepeater/SelectionModel was available so it does not use it. TreeView's multiselection does behave to get automatic selection of non-leaf nodes. Are you thinking of a different TreeView ? 

![image](https://user-images.githubusercontent.com/28935693/74337434-20702200-4d55-11ea-92d9-4859bfe77d4c.png)

**Comment by grokys** [comment]
I'm thinking of a `TreeView` like in the VS solution explorer or Window Explorer for example:

![image](https://user-images.githubusercontent.com/1775141/74338351-4bcd2e00-4da2-11ea-995b-3384f487750b.png)

Here, selecting `TeachingTip_InteractionTests` does not select the child nodes, and selecting both child nodes does not select the parent. I think this is the most common type of `TreeView`?

`SelectionModel`'s beahvior is that of a "CheckTreeView" (as I think it's known in some frameworks) - a `TreeView` with checkboxes.

**Comment by ranjeshj** [comment]
I see. Yes currently this is not exposed and the default is a tristate, but this seems like a reasonable thing to do. SelectionNode::IsSelected returns true/false  based on its selection and we also have SelectionNode::IsSelectedWithPartial which is what is currently used. Perhaps it just needs to be exposed though SelectionModel. SelectionChanged args will also need to differentiate between the two perhaps.

**Comment by grokys** [comment]
Thanks! I put together a quick hack in our Avalonia port which added a property `SelectOnlyLeafNodes` that when set to true enables the current "CheckTreeView" behavior, and when false works as a standard `TreeView` selection:

https://github.com/AvaloniaUI/Avalonia/commit/9b77a737e480bb48475c7786f90845f0be65401e
https://github.com/AvaloniaUI/Avalonia/commit/24a054137e6e08e8b572b34e3fc8163859a2fb94#diff-7adcd85c4a29ea22c4b1fb05de777d35

That seems to work from my initial messing around, but I suspect there's more to it. TBH by this point I'd given up in keeping a semblance of compatibility between WinUI and Avalonia and was just modifying stuff willy-nilly ;) However if possible I'd like to maintain some level of API compatibility between the two.

I'll revisit this soon and try to come up with a decent proposal. In the meantime if you have any feedback on the commits above, let me know!

**Comment by ranjeshj** [comment]
Thanks. Please keep these feedback issues coming. SelectionModel is still in preview and we can make api updates as needed. As you are already aware, most of the team is focused on WinUI3 so we won't be able to get to these until then.

**Comment by anawishnoff** [comment]
@grokys  Excited to see a proposal on this. As Ranjesh said, SelectionModel is slightly flexible as its still in preview so your feedback comes at a good time - this seems like a good thing to implement for the WinUI TreeView. I'll circle back when we may have more capacity to look into this!

**Comment by grokys** [comment]
Thanks @anawishnoff for the feedback. I've implemented this functionality in the Avalonia port of `SelectionModel` in these commits if anyone is interested:

https://github.com/AvaloniaUI/Avalonia/pull/3583/commits/96fbd6c531313dc0ab9fc3eeb7265a693b58e75f
https://github.com/AvaloniaUI/Avalonia/pull/3583/commits/f103df6a9a8dba4ae5880b1abee937a76b30b37c
